test/libsolidity/SolidityEndToEndTest.cpp

```diff
@@ -7382,6 +7382,33 @@ BOOST_AUTO_TEST_CASE(inline_assembly_function_access)
 	BOOST_CHECK(callContractFunction("x()") == encodeArgs(u256(10)));
 }
 
+BOOST_AUTO_TEST_CASE(inline_assembly_labels_with_stack_info)
+{
+	char const* sourceCode = R"(
+		contract C {
+			function f(uint y) {
+				assembly {
+					y 7
+					jump(fun)
+					exit[-1]:
+					y add
+					0 mstore
+					return(0, 0x20)
+					fun:
+					{
+						entry[a, b]:
+						a := div(a, b)
+						pop
```

Comment:
Actually, this might be the solution to all my problems.

The general form would be
`let (stack_element_production) := AssemblyItem (stack_element_consumption)`

```
let := sstore(1,2)
let x := sload(2)
```

We can even do this for labels:
```
let _, _ := ret 2 jump(f)

let return_value := ret(_, _) :
  stop

let r, arg := f(return_value):
let := jump(r)
let := pop(arg)
```

The scope of a variable is since it's produced till it's consumed.  Then we can also remove blocks.  Maybe this is for the even-lower level language.